File: minirextendr/R/status.R
# Status and introspection functions

#' Check if current project has miniextendr setup
#'
#' @param path Path to the R package root, or `"."` to use the current directory.
#' @return TRUE if the project appears to be a miniextendr package
#' @export
has_miniextendr <- function(path = ".") {
  with_project(path)
  is_miniextendr_package()
}

#' Show miniextendr setup status
#'
#' Displays which miniextendr files are present and which are missing
#' in the current project. For a comprehensive diagnostic, use
#' [miniextendr_doctor()] instead.
#'
#' @param path Path to the R package root, or `"."` to use the current directory.
#' @return Invisibly returns a list with present and missing files
#' @keywords internal
#' @export
miniextendr_status <- function(path = ".") {
  with_project(path)
  cli::cli_h1("miniextendr status")

  # Derive wrapper filename from actual package name
  pkg_name <- tryCatch(
    mx_desc_get_field("Package", file = usethis::proj_path("DESCRIPTION")),
    error = function(e) "miniextendr"
  )
  wrapper_file <- paste0("R/", pkg_name, "-wrappers.R")

  # Define expected files
  expected <- list(
    "Build System" = c(
      "configure.ac",
      "configure",
      "bootstrap.R",
      "cleanup",
      "cleanup.win",
      "cleanup.ucrt",
      "configure.win",
      "configure.ucrt",
      "tools/config.guess",
      "tools/config.sub"
    ),
    "Rust Project" = c(
      "src/rust/Cargo.toml",
      "src/rust/lib.rs",
      "src/rust/build.rs"
    ),
    "Source Files" = c(
      "src/Makevars.in",
      "src/stub.c",
      "src/cdylib-exports.def"
    ),
    "Generated Files" = c(
      "src/Makevars",
      wrapper_file
    )
  )

  present <- list()
  missing <- list()

  for (category in names(expected)) {
    cli::cli_h2(category)
    cat_present <- character()
    cat_missing <- character()

    for (file in expected[[category]]) {
      file_path <- usethis::proj_path(file)
      exists <- fs::file_exists(file_path) || fs::dir_exists(file_path)

      if (exists) {
        cli::cli_alert_success("{.path {file}}")
        cat_present <- c(cat_present, file)
      } else {
        cli::cli_alert_warning("{.path {file}} - missing")
        cat_missing <- c(cat_missing, file)
      }
    }

    present[[category]] <- cat_present
    missing[[category]] <- cat_missing
  }

  # Staleness check: compare .in templates vs generated files
  cli::cli_h2("Staleness")

  template_pairs <- list(
    c("src/Makevars.in", "src/Makevars")
  )

  stale <- character()
  for (pair in template_pairs) {
    tmpl_path <- usethis::proj_path(pair[1])
    gen_path <- usethis::proj_path(pair[2])
    if (fs::file_exists(tmpl_path) && fs::file_exists(gen_path)) {
      if (file.mtime(tmpl_path) > file.mtime(gen_path)) {
        cli::cli_alert_warning("{.path {pair[2]}} is stale (template {.path {pair[1]}} is newer)")
        stale <- c(stale, pair[2])
      }
    }
  }

  if (length(stale) == 0) {
    cli::cli_alert_success("All generated files up to date")
  } else {
    cli::cli_alert_info("Run {.code ./configure} to regenerate stale files")
  }

  # Summary
  total_present <- sum(lengths(present))
  total_missing <- sum(lengths(missing))
  total <- total_present + total_missing

  cli::cli_h2("Summary")
  cli::cli_alert_info("{total_present}/{total} files present")

  if (total_missing > 0) {
    cli::cli_alert_warning("{total_missing} files missing")
  }

  if (length(stale) > 0) {
    cli::cli_alert_warning("{length(stale)} stale generated file(s)")
  }

  invisible(list(present = present, missing = missing, stale = stale))
}

#' Validate miniextendr configuration
#'
#' Checks that the miniextendr setup is valid and ready to build.
#' For a comprehensive diagnostic, use [miniextendr_doctor()] instead.
#' For running `R CMD check`, see [miniextendr_check()].
#'
#' @param path Path to the R package root, or `"."` to use the current directory.
#' @return Invisibly returns TRUE if valid, otherwise shows warnings/errors
#' @keywords internal
#' @export
miniextendr_validate <- function(path = ".") {
  with_project(path)
  cli::cli_h1("miniextendr check")

  issues <- character()
  warnings <- character()

  # Check DESCRIPTION
  cli::cli_h2("DESCRIPTION")
  desc_path <- usethis::proj_path("DESCRIPTION")
  if (!fs::file_exists(desc_path)) {
    issues <- c(issues, "DESCRIPTION not found")
    cli::cli_alert_danger("DESCRIPTION not found")
  } else {
    # Check Config fields
    bootstrap <- mx_desc_get_field("Config/build/bootstrap", file = desc_path, default = "")
    if (bootstrap != "TRUE") {
      warnings <- c(warnings, "Config/build/bootstrap should be TRUE")
      cli::cli_alert_warning("Config/build/bootstrap not set to TRUE")
    } else {
      cli::cli_alert_success("Config/build/bootstrap = TRUE")
    }

    # Check SystemRequirements
    sys_req <- mx_desc_get_field("SystemRequirements", file = desc_path, default = "")
    if (!grepl("Rust", sys_req, ignore.case = TRUE)) {
      warnings <- c(warnings, "SystemRequirements should mention Rust")
      cli::cli_alert_warning("SystemRequirements doesn't mention Rust")
    } else {
      cli::cli_alert_success("SystemRequirements mentions Rust")
    }
  }

  # Check configure.ac
  cli::cli_h2("configure.ac")
  configure_ac <- usethis::proj_path("configure.ac")
  if (!fs::file_exists(configure_ac)) {
    issues <- c(issues, "configure.ac not found")
    cli::cli_alert_danger("configure.ac not found")
  } else {
    contents <- readLines(configure_ac, warn = FALSE)
    pkg_name <- get_package_name()

    # Check AC_INIT
    if (!any(grepl(paste0("AC_INIT.*", pkg_name), contents))) {
      warnings <- c(warnings, "AC_INIT doesn't match package name")
      cli::cli_alert_warning("AC_INIT package name may not match DESCRIPTION")
    } else {
      cli::cli_alert_success("AC_INIT package name matches")
    }
  }

  # Check Rust toolchain
  cli::cli_h2("Rust toolchain")
  tryCatch(
    {
      check_rust()
      rustc_version <- run_command("rustc", "--version")
      cli::cli_alert_success("Rust installed: {rustc_version}")
    },
    error = function(e) {
      issues <- c(issues, "Rust not found")
      cli::cli_alert_danger("Rust not found")
    }
  )

  # Check Cargo.toml declares miniextendr-api
  cli::cli_h2("Cargo.toml")
  cargo_toml_path <- usethis::proj_path("src", "rust", "Cargo.toml")
  if (!fs::file_exists(cargo_toml_path)) {
    issues <- c(issues, "src/rust/Cargo.toml not found")
    cli::cli_alert_danger("{.path src/rust/Cargo.toml} not found")
  } else {
    cargo_contents <- readLines(cargo_toml_path, warn = FALSE)
    if (any(grepl("miniextendr-api", cargo_contents, fixed = TRUE))) {
      cli::cli_alert_success("{.path src/rust/Cargo.toml} declares {.code miniextendr-api}")
    } else {
      issues <- c(issues, "src/rust/Cargo.toml is missing miniextendr-api dependency")
      cli::cli_alert_danger("{.path src/rust/Cargo.toml} is missing {.code miniextendr-api} dependency")
    }
  }

  # Summary
  cli::cli_h2("Result")
  if (length(issues) > 0) {
    cli::cli_alert_danger("{length(issues)} issue(s) found")
    for (issue in issues) {
      cli::cli_bullets(c("x" = issue))
    }
    invisible(FALSE)
  } else if (length(warnings) > 0) {
    cli::cli_alert_warning("{length(warnings)} warning(s)")
    cli::cli_alert_success("Package should build, but check warnings above")
    invisible(TRUE)
  } else {
    cli::cli_alert_success("All checks passed!")
    invisible(TRUE)
  }
}

File: miniextendr-cli/src/commands/status.rs
use std::path::Path;

use anyhow::Result;
use serde::Serialize;

use crate::bridge::{has_program, program_version};
use crate::cli::StatusCmd;
use crate::project::{MINIEXTENDR_CRATES, ProjectContext};

#[derive(Serialize)]
struct HasResult {
    has_miniextendr: bool,
    has_cargo_manifest: bool,
    has_description: bool,
    has_configure_ac: bool,
    has_configure: bool,
}

impl std::fmt::Display for HasResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "miniextendr: {}", yn(self.has_miniextendr))?;
        writeln!(f, "  src/rust/Cargo.toml: {}", yn(self.has_cargo_manifest))?;
        writeln!(f, "  DESCRIPTION: {}", yn(self.has_description))?;
        writeln!(f, "  configure.ac: {}", yn(self.has_configure_ac))?;
        write!(f, "  configure: {}", yn(self.has_configure))
    }
}

fn yn(b: bool) -> &'static str {
    if b { "yes" } else { "no" }
}

pub fn dispatch(cmd: &StatusCmd, ctx: &ProjectContext, _quiet: bool, json: bool) -> Result<()> {
    match cmd {
        StatusCmd::Has => status_has(ctx, json),
        StatusCmd::Show => status_show(ctx, json),
        StatusCmd::Validate => status_validate(ctx, json),
    }
}

/// Native — check if project has miniextendr.
fn status_has(ctx: &ProjectContext, json: bool) -> Result<()> {
    let result = HasResult {
        has_miniextendr: ctx.has_miniextendr(),
        has_cargo_manifest: ctx.cargo_manifest.is_some(),
        has_description: ctx.description.is_some(),
        has_configure_ac: ctx.configure_ac.is_some(),
        has_configure: ctx.configure.is_some(),
    };

    if json {
        println!("{}", serde_json::to_string_pretty(&result)?);
    } else {
        println!("{result}");
    }

    if !result.has_miniextendr {
        std::process::exit(1);
    }
    Ok(())
}

/// Native — show which miniextendr files are present/missing.
fn status_show(ctx: &ProjectContext, json: bool) -> Result<()> {
    let root = &ctx.root;

    // Derive wrapper filename from DESCRIPTION
    let pkg_name = read_description_field(root, "Package").unwrap_or_else(|| "miniextendr".into());
    let wrapper_file = format!("R/{pkg_name}-wrappers.R");

    let categories: &[(&str, &[&str])] = &[
        (
            "Build System",
            &[
                "configure.ac",
                "configure",
                "bootstrap.R",
                "cleanup",
                "cleanup.win",
                "cleanup.ucrt",
                "configure.win",
                "configure.ucrt",
                "tools/config.guess",
                "tools/config.sub",
            ],
        ),
        (
            "Rust Project",
            &[
                "src/rust/Cargo.toml",
                "src/rust/lib.rs",
                "src/rust/build.rs",
                "src/rust/cargo-config.toml.in",
            ],
        ),
        ("Source Templates", &["src/Makevars.in", "src/stub.c"]),
    ];

    // Build vendored crate paths dynamically from constant
    let vendor_paths: Vec<String> = MINIEXTENDR_CRATES
        .iter()
        .map(|c| format!("vendor/{c}"))
        .collect();
    let vendor_refs: Vec<&str> = vendor_paths.iter().map(|s| s.as_str()).collect();

    let categories: Vec<(&str, Vec<&str>)> = categories
        .iter()
        .map(|(name, files)| (*name, files.to_vec()))
        .chain(std::iter::once(("Vendored Crates", vendor_refs)))
        .collect();

    #[derive(Serialize)]
    struct StatusReport {
        present: Vec<String>,
        missing: Vec<String>,
        stale: Vec<String>,
    }

    let mut all_present = Vec::new();
    let mut all_missing = Vec::new();

    if !json {
        println!("=== miniextendr status ===\n");
    }

    for (category, files) in &categories {
        if !json {
            println!("-- {category} --");
        }
        for file in files {
            let path = root.join(file);
            let exists = path.exists();
            if exists {
                if !json {
                    println!("  [ok] {file}");
                }
                all_present.push(file.to_string());
            } else {
                if !json {
                    println!("  [--] {file}");
                }
                all_missing.push(file.to_string());
            }
        }
        if !json {
            println!();
        }
    }

    // Generated files (including dynamic wrapper)
    let generated = ["src/Makevars", &wrapper_file];
    if !json {
        println!("-- Generated Files --");
    }
    for file in &generated {
        let path = root.join(file);
        let exists = path.exists();
        if exists {
            if !json {
                println!("  [ok] {file}");
            }
            all_present.push(file.to_string());
        } else {
            if !json {
                println!("  [--] {file}");
            }
            all_missing.push(file.to_string());
        }
    }
    if !json {
        println!();
    }

    // Staleness check
    let template_pairs = [
        ("src/Makevars.in", "src/Makevars"),
        (
            "src/rust/cargo-config.toml.in",
            "src/rust/.cargo/config.toml",
        ),
    ];

    let mut stale = Vec::new();
    if !json {
        println!("-- Staleness --");
    }
    for (tmpl, generated) in &template_pairs {
        let tmpl_path = root.join(tmpl);
        let gen_path = root.join(generated);
        if tmpl_path.exists()
            && gen_path.exists()
            && let (Ok(tmpl_meta), Ok(gen_meta)) = (tmpl_path.metadata(), gen_path.metadata())
            && let (Ok(tmpl_time), Ok(gen_time)) = (tmpl_meta.modified(), gen_meta.modified())
            && tmpl_time > gen_time
        {
            if !json {
                println!("  [!!] {generated} is stale (template {tmpl} is newer)");
            }
            stale.push(generated.to_string());
        }
    }
    if stale.is_empty() && !json {
        println!("  All generated files up to date");
    }

    // Summary
    let total = all_present.len() + all_missing.len();
    if !json {
        println!("\n-- Summary --");
        println!("  {}/{total} files present", all_present.len());
        if !all_missing.is_empty() {
            println!("  {} files missing", all_missing.len());
        }
        if !stale.is_empty() {
            println!("  {} stale generated file(s)", stale.len());
        }
    } else {
        let report = StatusReport {
            present: all_present,
            missing: all_missing,
            stale,
        };
        println!("{}", serde_json::to_string_pretty(&report)?);
    }

    Ok(())
}

/// Native — validate miniextendr configuration.
fn status_validate(ctx: &ProjectContext, json: bool) -> Result<()> {
    let root = &ctx.root;

    #[derive(Serialize)]
    struct ValidateReport {
        pass: Vec<String>,
        warn: Vec<String>,
        fail: Vec<String>,
    }

    let mut pass = Vec::new();
    let mut warn = Vec::new();
    let mut fail = Vec::new();

    if !json {
        println!("=== miniextendr validate ===\n");
    }

    // DESCRIPTION checks
    if !json {
        println!("-- DESCRIPTION --");
    }
    let desc_path = root.join("DESCRIPTION");
    if !desc_path.exists() {
        fail.push("DESCRIPTION not found".into());
        if !json {
            println!("  [FAIL] DESCRIPTION not found");
        }
    } else {
        let bootstrap = read_description_field(root, "Config/build/bootstrap").unwrap_or_default();
        if bootstrap == "TRUE" {
            pass.push("Config/build/bootstrap = TRUE".into());
            if !json {
                println!("  [ok] Config/build/bootstrap = TRUE");
            }
        } else {
            warn.push("Config/build/bootstrap should be TRUE".into());
            if !json {
                println!("  [!!] Config/build/bootstrap not set to TRUE");
            }
        }

        let sys_req = read_description_field(root, "SystemRequirements").unwrap_or_default();
        if sys_req.to_lowercase().contains("rust") {
            pass.push("SystemRequirements mentions Rust".into());
            if !json {
                println!("  [ok] SystemRequirements mentions Rust");
            }
        } else {
            warn.push("SystemRequirements should mention Rust".into());
            if !json {
                println!("  [!!] SystemRequirements doesn't mention Rust");
            }
        }
    }

    // configure.ac
    if !json {
        println!("\n-- configure.ac --");
    }
    if ctx.configure_ac.is_none() {
        fail.push("configure.ac not found".into());
        if !json {
            println!("  [FAIL] configure.ac not found");
        }
    } else {
        pass.push("configure.ac present".into());
        if !json {
            println!("  [ok] configure.ac present");
        }
    }

    // Rust toolchain
    if !json {
        println!("\n-- Rust toolchain --");
    }
    if has_program("rustc") {
        let version = program_version("rustc").unwrap_or("unknown".into());
        pass.push(format!("Rust installed: {version}"));
        if !json {
            println!("  [ok] Rust: {version}");
        }
    } else {
        fail.push("Rust not found - install from https://rustup.rs".into());
        if !json {
            println!("  [FAIL] Rust not found");
        }
    }

    if has_program("cargo") {
        pass.push("cargo available".into());
        if !json {
            println!("  [ok] cargo available");
        }
    } else {
        fail.push("cargo not found".into());
        if !json {
            println!("  [FAIL] cargo not found");
        }
    }

    // Vendored crates
    if !json {
        println!("\n-- Vendored crates --");
    }
    for krate in MINIEXTENDR_CRATES {
        if root.join("vendor").join(krate).is_dir() {
            pass.push(format!("{krate} vendored"));
            if !json {
                println!("  [ok] {krate}");
            }
        } else {
            warn.push(format!("{krate} not vendored"));
            if !json {
                println!("  [!!] {krate} not vendored");
            }
        }
    }

    // Summary
    if !json {
        println!("\n-- Result --");
        println!("  {} passed", pass.len());
        if !warn.is_empty() {
            println!("  {} warning(s)", warn.len());
        }
        if !fail.is_empty() {
            println!("  {} failure(s)", fail.len());
            for f in &fail {
                println!("  x {f}");
            }
        }
        if fail.is_empty() && warn.is_empty() {
            println!("  All checks passed!");
        }
    } else {
        let report = ValidateReport { pass, warn, fail };
        println!("{}", serde_json::to_string_pretty(&report)?);
    }

    Ok(())
}

/// Read a field from DESCRIPTION (simple DCF parser).
fn read_description_field(root: &Path, field: &str) -> Option<String> {
    let desc_path = root.join("DESCRIPTION");
    let content = std::fs::read_to_string(desc_path).ok()?;
    let prefix = format!("{field}:");

    for line in content.lines() {
        if line.starts_with(&prefix) {
            return Some(line[prefix.len()..].trim().to_string());
        }
    }
    None
}
